### ISSUE TITLE
Directus tries to refresh with expired refresh-token before every public SDK call

### ISSUE THREAD

**Issue by dataexcess** [directus/directus#10866]
### Preflight Checklist

- [X] I have completed all [Troubleshooting Steps](https://docs.directus.io/getting-started/support/#troubleshooting-steps).
- [X] I'm on [the latest version of Directus](https://github.com/directus/directus/releases).
- [X] There's [no other issue](https://github.com/directus/directus/issues?q=is%3Aissue) that already describes my problem.

### Describe the Bug

Hello there!

My website offers two ways to enter, one asking the user to login (calling the /login method etc.) and another 'anonymous' method, which displays the website in a more limited manner.

I have the following problem: when a user successfully registers and then logs in to the website, all is good and well. Now when the user tries to re-enter the website after the max lifetime of the refresh-token in the cookie is exceeded. The user catches the no-refresh token error and displays again the enter-panel of the website prompting the user to either 'login' or 'enter anonymously'. When the user chooses 'enter anonymously' all the methods made to the SDK still excpect the user to be logged in somehow and so it tries to call the token-refresh method which fails since there is no refresh token present in the cookie. As a result my website becomes dysfunctional and I have to manually remove the cookie in the browser preferences to be able to continue. 

1. how can I tell Directus to stop trying to refresh the token?
2. I cannot call the logout method since it also needs the access/refresh-token.
3. why is Directus trying to refresh the token in the first place for endpoints that are completely public with all access (not a restricted method such as me/read, but just a public collection/read call)

Please help me fix this problem.
Thank you!

### To Reproduce

1. set server refresh_token_life_time to "60s"
2. login
3. wait more than one minute
4. refresh the page
5. calll me/read method which causes a 'refresh_token missing'
6. call any other completely unrestricted public method.... everything fails

### Errors Shown

400 (Bad Request)
"message":""refresh_token" is required in either the JSON payload or Cookie".

### What version of Directus are you using?

9.0.0-rc.83

### What version of Node.js are you using?

v15.5.1

### What database are you using?

Postgres

### What browser are you using?

Chrome

### What operating system are you using?

Ubuntu

### How are you deploying Directus?

nginx

**Comment by BartTactick** [comment]
I experience the same issue. Basicly after your refreshtoken expires, you cannot access any content that the role "Public" could/should.

**Comment by azrikahar** [comment]
Would you mind updating to the latest version of Directus and try again? There are quite many substantial changes to both auth & SDK since rc.83 until now. If the problem still persists on the latest version, we can reopen and continue investigating 👍

**Comment by dataexcess** [comment]
> Would you mind updating to the latest version of Directus and try again? There are quite many substantial changes to both auth & SDK since rc.83 until now. If the problem still persists on the latest version, we can reopen and continue investigating 👍

Just updated to 9.4 and still the same bug!

**Comment by rijkvanzanten** [comment]
@dataexcess have you seen the `autoRefresh` configuration flag?

https://docs.directus.io/reference/sdk/#constructor

**Comment by dataexcess** [comment]
> @dataexcess have you seen the `autoRefresh` configuration flag?
> 
> https://docs.directus.io/reference/sdk/#constructor

I am not initialising the SDK with any customization so I assume the default is on meaning it automatically refreshes the accessToken with the refreshToken.

But how does this relate to my issue? Maybe I'm not getting your point sorry..

**Comment by rijkvanzanten** [comment]
> But how does this relate to my issue? Maybe I'm not getting your point sorry..

Because I think it's the answer to the questions you posted:

> how can I tell Directus to stop trying to refresh the token?

By setting autoRefresh to false, as that will prevent the SDK from automatically trying to refresh the token

> I cannot call the logout method since it also needs the access/refresh-token.

The logout endpoint will stop the current session, which indeed requires you to submit what session to stop through the refresh token.

> why is Directus trying to refresh the token in the first place for endpoints that are completely public with all access (not a restricted method such as me/read, but just a public collection/read call)

Because the SDK is configured to do so 🙂

**Comment by dataexcess** [comment]
> > But how does this relate to my issue? Maybe I'm not getting your point sorry..
> 
> Because I think it's the answer to the questions you posted:
> 
> > how can I tell Directus to stop trying to refresh the token?
> 
> By setting autoRefresh to false, as that will prevent the SDK from automatically trying to refresh the token
> 
> > I cannot call the logout method since it also needs the access/refresh-token.
> 
> The logout endpoint will stop the current session, which indeed requires you to submit what session to stop through the refresh token.
> 
> > why is Directus trying to refresh the token in the first place for endpoints that are completely public with all access (not a restricted method such as me/read, but just a public collection/read call)
> 
> Because the SDK is configured to do so 🙂

Hehe I really think all your answers make complete sense, and I really would love to keep enjoying the auto refreshing of tokens that the SDK offers.
But can't you agree that I'm in a weird situation here? How can I fix this? 

Why is it that, if I empty the cookie, call a public unrestricted method, the SDK does not throw any 'no refresh-token' present errors, but once there was a refresh-token, and then at a later point there is nomore refresh-token since it is expired, the same public method does throw an error...

**Comment by rijkvanzanten** [comment]
> Why is it that, if I empty the cookie, call a public unrestricted method, the SDK does not throw any 'no refresh-token' present errors, but once there was a refresh-token, and then at a later point there is nomore refresh-token since it is expired, the same public method does throw an error...

I'm assuming it's because the _access_ token was still being used, even though you removed the refresh token. Don't know for sure though! I'd have to see your exact implementation to answer that properly

**Comment by dataexcess** [comment]
Yes exactly! It must be something like that. So that is my question, how can I tell Directus to stop using the access_token? I am just using the very default initialisation SDK and letting Directus do all the access/refresh-token stuff, so I have no idea, should I manually do this?

**Comment by rijkvanzanten** [comment]
If you're logged in with the SDK, I would personally just call `logout()` and let the SDK figure it out from there. The refresh token persists in the cookie until you call logout manually, and (by default), the access token only exists in the browser memory for the lifespan of the access token.

**Comment by dataexcess** [comment]
> The refresh token persists in the cookie until you call logout manually

I don't understand? It does not when the refresh-token has an expiration date a.k.a REFRESH_TOKEN_TTL. Sure I could just put the REFRESH_TOKEN_TTL to some huge number and keep the user logged in forever, but I was hoping for a better solution :p

**Comment by dataexcess** [comment]
What I would do, or what I would expect, is for Directus to catch the refresh-token expired bug internally and remove the access-token or any other flags it has internally that it should refresh the token before every SDK call.

**Comment by rijkvanzanten** [comment]
@dataexcess Can you share a code reproduction where this is happening? I'm having some trouble following along with what your exact use case is and where its going wrong

**Comment by dataexcess** [comment]
I am logged out (no cookie) and I call the following method

`        let role = await API.roles.readMany({ filter: { name: { '_eq': 'Visitor' } } })
        .catch((err) => {
            console.error(err);
        })

        let visitorRole = role.data[0].id
        console.log("visitor role: " + visitorRole)`

which works. 
Then I login, I wait til the refresh_token life time expires... I refresh and I try the same method again and it fails

**Comment by rijkvanzanten** [comment]
Is your access token TTL longer than your refresh token TTL? @joselcvarela does the SDK auto refresh on 401 errors as well as time based?

**Comment by dataexcess** [comment]
> Is your access token TTL longer than your refresh token TTL? @joselcvarela does the SDK auto refresh on 401 errors as well as time based?

No. These are the values I'm testing with:
ACCESS_TOKEN_TTL="30s"
REFRESH_TOKEN_TTL="60s"

**Comment by dataexcess** [comment]
@rijkvanzanten Can you confirm that this is indeed an issue? Maybe I have the first website using Directus which offers a non-logged-in/anonymous version to be used by a user after refresh-token expires. hehe.

**Comment by joselcvarela** [comment]
Although you mentioned you tested directus with 9.0.0-rc.83 and 9.4 you didn't mention what version you are using of `@directus/sdk`? Can you do

```sh
npm i @directus/sdk@latest
```

and check if the problem occurs?

**Comment by dataexcess** [comment]
> 

these are my dependencies in my package-lock.json file:

"dependencies": {
        "@directus/sdk": "^9.0.0-rc.83",
        "@directus/sdk-js": "^9.0.0-rc.53",
        etc.

**Comment by dataexcess** [comment]
> Although you mentioned you tested directus with 9.0.0-rc.83 and 9.4 you didn't mention what version you are using of `@directus/sdk`? Can you do
> 
> ```shell
> npm i @directus/sdk@latest
> ```
> 
> and check if the problem occurs?

Ok I called your command, and now I'm on a later version, but now my directus loggin method does not work anymore :/
"@directus/sdk": "^9.4.3",
"@directus/sdk-js": "^9.0.0-rc.53",

When calling 
```
        let loginuser = await API.auth.login({
            email: email,
            password: password
        })
        .catch( error => {
            console.log("error when logging in: " + error)
        })
```
it prints: error when logging in: Error: An unexpected error occurred.

**Comment by rijkvanzanten** [comment]
`@directus/sdk` is the correct one. Make sure to use the version of the SDK that's made to work with your version of `directus` itself. If you use Directus `9.4.2`, make sure to use `@directus/sdk` version `9.4.2`. 

Also, I would highly recommend staying up to date with the latest versions, as we patch a lot of these issues over time. At least ensure to be on the production release (9.0.0+) 🙂

**Comment by joselcvarela** [comment]
The error now comes from Directus instance and not from the SDK.

Try to run your directus with this command
`DEBUG=* npx directus start`

And then, try to make that request via curl or other client to make HTTP requests
```
POST /auth/login

{ "email": "EMAIL", "password": "PASSWORD" }
```

Check the console to see if you can find any errors.

Otherwise I suggest to run with inspector enabled and debug your instance manually:
`node --inspect node_modules/.bin/directus start`

**Comment by dataexcess** [comment]
Hello everyone,

Finally got back to this issue. So I updated my directus server instance to 9.5.0 as well as my directus-sdk to 9.5.0.
The logging in problem persists. and I cannot even log in to the directus app directly, same issue.
I ran the server instance with the "DEBUG=* npx directus start" command.
This seems to be the issue:

```
10:40:39 🚨 "expiresIn" should be a number of seconds or string representing a timespan eg: "1d", "20h", 60 
Error: "expiresIn" should be a number of seconds or string representing a timespan eg: "1d", "20h", 60
    at Object.module.exports [as sign] (/home/dataexcess/SubNeu_backend/node_modules/jsonwebtoken/sign.js:175:22)
    at AuthenticationService.login (/home/dataexcess/SubNeu_backend/node_modules/directus/dist/services/authentication.js:139:52)
    at async /home/dataexcess/SubNeu_backend/node_modules/directus/dist/auth/drivers/local.js:67:56
10:40:39 ✨ request errored POST 500 /auth/login 43ms
```

How can I fix this please? (without having access to the app)

**Comment by dataexcess** [comment]
OOOOW snap!
Ok I naively assumed I could add a comment at the end of my .env file variables. seems like removing those comments fixed it 😅
learning progress. . .

**Comment by dataexcess** [comment]
Ok so back to the question at hand.
Updating to the latest Directus version did change some things (for the better).

I set my ACCESS_TOKEN_TTL to 2 minutes and my REFRESH_TOKEN_TTL to 4 min.

I log in with my credentials and enter my website with unrestriced access, good.
in the time period before 2 minutes passed, i can refresh my website and my user stays logged in (access token is used), good.
after 2 minutes have passed, I try to refresh and I am locked out of the website: when I try to do the "users.me" call it says :

`Error: Token expired.`

this is bad, I thought the refresh_token would refresh the access_token automatically? Has this been changed somewhere in the new version? If after this error pops up (and the login screen is displayed) I enter my website anonymously instead of logging in, I get again the same bug when accessing public api endpoints:

```
GET https://subneu.live/api/files?filter=%7B%22filename_download%22:%7B%22_eq%22:%22TreeSound.mp3%22%7D%7D 401 (Unauthorized)
Network.js?32ba:98 Error: Token expired.
Network.js?32ba:102 Uncaught (in promise) TypeError: Cannot read properties of undefined (reading 'data')
```

If I wait till we pass the timeframe of 4 minutes, and I refresh the page, and the users.me method is called, I get what I would expect when the refresh_token is expired:

a simple 
`Error: Invalid user credentials.
`
instead of 
`Error: "refresh_token" is required in either the JSON payload or Cookie
`
This is good! I can then choose to enter anonymously or log in again, and both are working without any issues yeey!

So, long story short: it seems for my use case the refresh_token weird behaviour is fixed. but now it is the access_token making trouble... maybe I need to specify somewhere to use the refresh_token to refresh the access_token?

Thank you!

**Comment by licitdev** [comment]
To refresh the access token:

If you're using the JS SDK
https://docs.directus.io/reference/sdk/#refresh-auth-token

Alternatively, you may perform a POST request to refresh the token https://docs.directus.io/reference/authentication/#refresh

If you wait past 4 minutes, your refresh token would have expired, hence you will not be able to get the access token. You have to refresh the access token before or when it is expired, while the refresh token is still valid.

**Comment by dataexcess** [comment]
> To refresh the access token:
> 
> If you're using the JS SDK https://docs.directus.io/reference/sdk/#refresh-auth-token
> 
> Alternatively, you may perform a POST request to refresh the token https://docs.directus.io/reference/authentication/#refresh
> 
> If you wait past 4 minutes, your refresh token would have expired, hence you will not be able to get the access token. You have to refresh the access token before or when it is expired, while the refresh token is still valid.

But isn't the access_token refreshing supposed to happen automatically? I am not specifying an y initialisation stuff when creating the SDK instance so it should use the default values where autorefresh is true.

**Comment by licitdev** [comment]
@dataexcess Yes, the refreshing should be automatic, merged in #9777. 
Just to confirm, is the `refresh_token` successfully saved as a cookie?

Please refer to discussion #9720. There are some details and workarounds on refresh token cookie's domain and also on secure cookies. Thank you.

**Comment by dataexcess** [comment]
I

> @dataexcess Yes, the refreshing should be automatic, merged in #9777. Just to confirm, is the `refresh_token` successfully saved as a cookie?
> 
> Please refer to discussion #9720. There are some details and workarounds on refresh token cookie's domain and also on secure cookies. Thank you.

I am confused. Is this error "Token expired" referring to the access_token or the refresh_token?

When I test, I can see the cookie 'directus_refresh_token' in the dev tools, also it changes its value "expires/max age" every 2/4 minutes.. (so I am assuming this is when the access_token is expired, and is being refreshed by the refresh_token, and it expands its own lifetime as well?

When I close the website, and I wait til I pass the value 'expires/max-age' of my refresh_token cookie, and I reopen the website and I call me.read(), then the 'token expired' comes up.

Which is fine, if that is its purpose, to signal that the refresh_token is expired. the problem is that I cannot call any public method once I received this error (by entering my website in anonymous mode)

**Comment by dataexcess** [comment]
When I am at this stage... the "token expired" error being shown, and I wait a few more minutes, eventually my local storage auth_session object is deleted. When I then refresh the page, I simply get a "Invalid user credentials." error and I can log in to the website, or choose the anonymous route and all methods work.

I am super confused about all of this, sorry. Maybe I'm simpl;y not understanding it, but I would simply expect for the access_token to keep being refreshed behind the scenes, unless I close the website and wait longer than the refresh_token lifetime. In that case I would simply be greeted with a "Invalid user credentials." error and I could enter my website again by logging in or going anonymous without a problem. 

But it seems that somewhere in the time when the access_token lifetime has passed and the auth_token is still in the localStorage that this unwanted "token expired" error is causing trouble.

**Comment by dataexcess** [comment]
Ok, There was definitely something weird going on and I might have catched an edge-case. But after some more extensive testing, I believe everything is working as it should.

Sorry for the hassle/confusion!

**Comment by adamkhan** [comment]
I am having the same issue running v10.8.2. @dataexcess, how did this resolve itself? The only way I can access my site after my 1h has expired is to manually delete the cookies, otherwise every query just gives me the error "TOKEN_EXPIRED".

**Comment by rijkvanzanten** [comment]
@adamkhan This issue is close to 2 years old. Please open a new issue, as the root cause and solve is most likely completely unrelated 👍🏻
